core/drivers/versal_pm.c

```diff
@@ -0,0 +1,147 @@
+#include <kernel/cache_helpers.h>
+#include <kernel/delay.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <tee/cache.h>
+#include <tee_api_types.h>
+#include <utee_defines.h>
+
+#define PAYLOAD_ARG_CNT		6
+#define PM_SIP_SVC		0xc2000000
+
+/* PM API ids */
+#define PM_GET_API_VERSION		1
+#define PM_GET_DEVICE_STATUS		3
+#define PM_GET_OP_CHARACTERISTIC	4
+#define PM_REGISTER_NOTIFIER		5
+#define PM_REQ_SUSPEND			6
+#define PM_SELF_SUSPEND			7
+#define PM_FORCE_POWERDOWN		8
+#define PM_ABORT_SUSPEND		9
+#define PM_REQ_WAKEUP			10
+#define PM_SET_WAKEUP_SOURCE		11
+#define PM_SYSTEM_SHUTDOWN		12
+#define PM_REQUEST_DEVICE		13
+#define PM_RELEASE_DEVICE		14
+#define PM_SET_REQUIREMENT		15
+#define PM_SET_MAX_LATENCY		16
+#define PM_RESET_ASSERT			17
+#define PM_RESET_GET_STATUS		18
+#define PM_INIT_FINALIZE		21
+#define PM_GET_CHIPID			24
+#define	PM_PINCTRL_REQUEST		28
+#define	PM_PINCTRL_RELEASE		29
+#define	PM_PINCTRL_GET_FUNCTION		30
+#define	PM_PINCTRL_SET_FUNCTION		31
+#define	PM_PINCTRL_CONFIG_PARAM_GET	32
+#define	PM_PINCTRL_CONFIG_PARAM_SET	33
+#define PM_IOCTL			34
+#define PM_QUERY_DATA			35
+#define PM_CLOCK_ENABLE			36
+#define PM_CLOCK_DISABLE		37
+#define PM_CLOCK_GETSTATE		38
+#define PM_CLOCK_SETDIVIDER		39
+#define PM_CLOCK_GETDIVIDER		40
+#define PM_CLOCK_SETRATE		41
+#define PM_CLOCK_GETRATE		42
+#define PM_CLOCK_SETPARENT		43
+#define PM_CLOCK_GETPARENT		44
+#define PM_PLL_SET_PARAMETER		48
+#define PM_PLL_GET_PARAMETER		49
+#define PM_PLL_SET_MODE			50
+#define PM_PLL_GET_MODE			51
+#define PM_FEATURE_CHECK		63
+#define PM_LOAD_PDI			0x701
+
+/* PDI sources */
+#define PDI_SRC_JTAG		0x0
+#define PDI_SRC_QSPI24		0x1
+#define PDI_SRC_QSPI32		0x2
+#define PDI_SRC_SD0		0x3
+#define PDI_SRC_EMMC0		0x4
+#define PDI_SRC_SD1		0x5
+#define PDI_SRC_EMMC1		0x6
+#define PDI_SRC_USB		0x7
+#define PDI_SRC_OSPI		0x8
+#define PDI_SRC_SBI		0x9
+#define PDI_SRC_SMAP		0xA
+#define PDI_SRC_PCIE		0xB
+#define PDI_SRC_SD1_LS		0xE
+#define PDI_SRC_DDR		0xF
+
+struct versal_sip_payload {
+	uint32_t data[PAYLOAD_ARG_CNT];
+};
+
+static uint32_t versal_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				struct versal_sip_payload *payload)
+{
+	struct thread_smc_args args = {
+		.a0 = PM_SIP_SVC | pm_api_id,
+		.a1 = reg_pair_to_64(arg1, arg0),
+		.a2 = reg_pair_to_64(arg3, arg2),
+	};
+
+	thread_smccc(&args);
+
+	if (payload) {
+		reg_pair_from_64(args.a0, &payload->data[1], &payload->data[0]);
+		reg_pair_from_64(args.a1, &payload->data[3], &payload->data[2]);
+		reg_pair_from_64(args.a2, &payload->data[5], &payload->data[4]);
+	}
+
+	/* allow the PLM to output its debug information */
+	if (IS_ENABLED(CFG_VERSAL_TRACE_PLM))
+		mdelay(500);
+
+	return args.a0;
+}
+
+TEE_Result versal_write_fpga(paddr_t pa)
+{
+	struct ipi_cmd cmd = { };
+
+	if (IS_ENABLED(CFG_VERSAL_FPGA_PM_DRIVER))
+		return versal_sip_call(PM_LOAD_PDI, PDI_SRC_DDR, pa, pa >> 32,
```

Comment:
After a bit of searching I found this comment in TF-A where the `PM_LOAD_PDI` call is handled:
```
/**
 * eemi_for_compatibility() - EEMI calls handler for deprecated calls
 *
 * @return - If EEMI API found then, uintptr_t type address, else 0
 *
 * Some EEMI API's use case needs to be changed in Linux driver, so they
 * can take advantage of common EEMI handler in TF-A. As of now the old
 * implementation of these APIs are required to maintain backward compatibility
 * until their use case in linux driver changes.
 */
```
It seems you're calling a deprecated function. Any chance of using a non-deprecated function instead?

By the way, it would be nice with a reference to the ABI and which class of platform can expect this ABI to be available.

Reply:
right. I could just remove it - just thought it might be useful as a backup. 

The preferred call is the one right under it (direct to the PLM instead of going to the TF-A)

Reply:
Backup sounds almost like it's not going to be used or tested regularly.
If it works for you without this perhaps it's better to remove it.